### DISCUSSION TITLE
자바스크립트 딥 다이브 1 ~ 9장 팔요한 것 정리

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#32] [ALL STUFF]
자바스크립트는 이제껏 제가 주로 다뤄웠던 클래스 기반의 객체지향은 아니지만 객체지향언어가 맞습니다. 다만 클래스 기반이 아닌 프로토 타입 기반이라는 점이 다를뿐입니다. (ES6 부터는 클래스 개념도 추가되었습니다.) 여기서 말하는 객체지향언어는 일정한 객체를 만들기위해 청사진을 만들어 놓고 그 청사진을 기반으로 동일한 성질을 갖는 객체들을 생성하고 또 이를 활용하는 프로그래밍 언어를 말합니다.

![20240610_203239.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/137bc349-0336-4d21-a038-11c14da32842)

 자바스크립트에서 원시 타입의 값들은 변경이 불가능합니다. 이 말이 의미하는것은 특정 변수에 값을 할당하면 메모리에 해당 값과 그 값을 저장한 메모리 주소를 변수가 바라보도록 하게 되는데 이때 값을 변경한다면 메모리에 저장된 값을 변경하는것이 아닌 새로운 메모리에 값을 저장하고 새로운 메모리 주소를 변수가 바라보도록 라는 것입니다.
 만약 변수를 선언만 한다하더라도 메모리에는 undefined 가 저장되게 됩니다. 심지어 선언과 동시에 값을 할당한다하더라도 동일한 절차를 밟아 undefined 가 저장된 뒤에 새로운 변수가 저장된 메모리를 바라보게 됩니다. 이는 메모리에 저장된 값을 직접 건드리지 않음으로써 안정성을 확보하기 위함입니다. 또한 변수가 바라보지 않는 메모리 주소의 값들은 가비지 컬렉터가 정리를 하게 됩니다.

![20240610_203529.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/1f01e711-6813-47d4-a2cf-36d71ee65705)

![20240610_203535.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/f9162bdf-965f-49a6-8831-dfdfcba4bdb0)

 자바스크립트에는 정수와 실수의 구분이 없습니다. 이 말이 의미하는것은 모든 수를 실수로써 처리한다는 의미입니다. 따라서 정수로 보이는 두 수를 나눈다면 실수가 나올수도 있습니다.

![20240610_203737.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/0ba9981e-a775-4feb-8a25-1f93907e4759)

![20240610_203904.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/3bde9643-2012-4bf0-b996-a5f7f1d1d0b6)

 자바스크립트에서 숫자 연산을 할 시에 피연산자중 숫자 연산이 불가능한 값이 있다면 (예를들면 number 로 형변환이 불가능한 문자열) NaN 이라는 값이 나오게 됩니다. 그런데 이 NaN 은 자기 자신과 비교연산을 할때 문제가 있습니다. NaN === NaN 을 한다면 결과값이 false 가 됩니다. 그래서 NaN 여부를 알고 싶다면 Number.isNaN 을 써야 합니다.

 자바스크립트에는 +0 과 -0 이 따로 존재합니다만 두 값을 비교하면 동일하다고 취급됩니다.

 논리연산자를 쓸때 드 모르간 법칙을 이용하면 가독성을 높일수도 있습니다.
!(x ||  y) === !x && !y
!(x && y) === !x || !y

 이터레이블한 객체를 이용할때 for ... of 로 각각의 요소에 접근할 수 있습니다. 일반 객체를 이용할때 for ... in 을 이용하면 각각의 프로퍼티에 접근할 수 있습니다.

  falsy 한 값들중 0과 '' 는 객체로 평가되는 경우도 존재합니다. (추후 배움.)

![20240610_203956.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/a51467d8-675e-4643-ab7e-c4a8dec9b30f)

 옵셔널 체이닝 연산자는 대상이 null 또는 undefined 가 아니면 멈추지 않고 계속 참조합니다.

 널리쉬 연산자중 || 와 ?? 의 차이점은 || 는 모든 falsy 값들을 처리하지만 ?? 는 null 과 undefined 만 처리합니다.

![20240610_204015.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/e61f176e-2dd7-4904-886b-aa71160890b1)

 레이블문은 문을 breakable 하게 만들어주어서 원하는 시점에 break 를 함으로써 스코프를 빠져 나올수 있도록 합니다.
test: {
  .
  .
  .
  break test;
}

 특정 값 앞에 !! 를 붙임으로써 불린타입으로 형변환할 수 있습니다. 만약 ! 를 붙인다면 형변환과 동시에 값을 반전 시킵니다.

![20240610_204042.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/ca1b4c7f-a457-4779-8c54-7e26e5e65663)

![20240610_204722.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/689a7e4f-036b-4790-935f-2b29ef1e89c3)

 타입변환에는 여러 방식이 있지만 생성자를 이용하여 형변환을 할 수도 있습니다. 이 경우 new 키워드는 생략합니다.
String(100);

![20240610_204057.jpg](https://github.com/honaseung/react-deep-dive/assets/53066347/9edca604-a1b7-4780-bf26-3760a4655a90)

 const test = 'cat' || 'dog'; 이 있을때 test 는cat 이 됩니다. 이는 nullish 연산자로 볼 수도 있어서 금방 이해가 갑니다.
 하지만 주의해야할건 const test ='cat' && 'dog'; 일 경우에는 test 가 dog 이 된다는 점입니다. 논리곱과 논리합 연산자는 피연산자를 형변환하지 않고 그대로 반환합니다. 이는 논리곱과 합에서 뒤의 피연산자를 고려할 필요가 없는 경우 단축 평가가 일어나기 때문입니다.

 리터럴이란 사람이 이해할 수 있도록 문자 또는 약속된 기호로 값을 생성하는 기법입니다. 따라서 자바스크립트 엔진를 한번 거쳐서 값을 생성합니다.

**Comment by chaminju927** [comment]
심지어 선언과 동시에 값을 할당한다하더라도 동일한 절차를 밟아 undefined 가 저장된 뒤에 새로운 변수가 저장된 메모리를 바라보게 됩니다.  -> 이 부분에서 저는 변수 선언과 동시에 할당을 하면 값이 저장된 주소를 바로 변수가 바라보는 것으로 알고 있었네요 ㅠㅠ 혹시 책에 어느부분에 나와있는 내용인가요 ?

**Reply by honaseung** [reply]
44 쪽의 예제 4-9 밑에 있는 내용입니다.

**Comment by chaminju927** [comment]
마지막 리터럴에 관한 내용에서 자바스크립트엔진을 한 번 거쳐서 값을 생성한다 -> 이부분이 잘 이해가 가지 않는데 조금 더 자세히 알려주실 수 있을까요?

**Reply by honaseung** [reply]
예를 들면 숫자 9 가 있을때, 이 숫자 9 는 아라비아 숫자로써 사람이 한 눈에 알아 볼 수 있는 값입니다. 하지만 컴퓨터는 이진법 밖에 사용할 수 없습니다. 따라서 이 숫자 9 가 컴퓨터가 알아 듣기 위해서는 이진법으로 변환하는 작업이 필요합니다. 여기서 변환하는 작업을 자바스크립트 엔진을 한번 거친다고 표현하였습니다.

**Comment by chaminju927** [comment]
nullish가 정확히 무엇을 의미하는 건가요?

**Reply by honaseung** [reply]
정확한 표현으로는 nullish 병합 연산자(Nullish coalescing operator) 라고 합니다. nullish 라는 것은 js 에서 falsy 한 값들 중에서 null 과 undefined 만을 가르킵니다.  다른 falsy 한 값들인 빈문자열(''), 숫자 영(0) 등등을 제외하여 가르키는 값들입니다. 이는 빈문자열과 숫자 등등을 실제로 의미 있는 값으로 바라볼 경우에 사용한다면 유용하게 이용할 수 있는 연산자 입니다.
